lib/pescarte/modulo_pesquisa/relatorio_compiler.ex

```diff
@@ -0,0 +1,41 @@
+defmodule Pescarte.RelatorioCompiler do
+  alias Pescarte.ModuloPesquisa.Repository
+  alias PescarteWeb.RelatorioHTML
+
+  def compilar_relatorios(relatorios_selecionados) do
+    relatorios = Enum.map(relatorios_selecionados, &find_relatorio/1)
+    htmls = Enum.map(relatorios, &RelatorioHTML.content/1)
+    pdfs = Enum.reduce(htmls, [], fn html, acc -> [{:html, html} | acc] end)
+
+    with {:ok, pdf_compilado} <- ChromicPDF.print_to_pdf(pdfs),
+         zip_file <- criar_zip_file(pdf_compilado) do
+      elem(zip_file, 1)
+    end
+  end
+
+  def gerar_pdf(relatorio_id) do
+    relatorio = find_relatorio(relatorio_id)
+
+    with {:ok, pdf} <-
+           [content: RelatorioHTML.content(relatorio), size: :a4]
+           |> ChromicPDF.Template.source_and_options()
+           |> ChromicPDF.print_to_pdf() do
+      Base.decode64(pdf)
+    end
+  end
+
+  defp find_relatorio(relatorio_id) do
+    Repository.fetch_relatorio_pesquisa_by_id(relatorio_id)
+  end
+
+  defp criar_zip_file(pdf) do
+    with {:ok, zip_file} <-
+           :zip.create(
+             "relatorios_compilados.zip",
+             [{~c"relatorios_compilados.pdf", Base.decode64!(pdf)}],
+             [:memory]
+           ) do
+      zip_file
+    end
+  end
```

Comment:
acho que o ideal seria ficar assim:
```suggestion
  @zip_filename “relatorio_final.zip"
  @zipped_filename “relatorio_final.pdf"
  
  defp criar_zip_file(pdf) do
    with {:ok, pdf} <- Base.decode64(pdf),
         files = [{String.to_charlist(@zipped_filename), pdf}],
         {:ok, {_, binary}} <- :zip.create(@zip_filename, files [:memory]) do
       {:ok, @zip_filename, binary}
    end
  end
```

dessa forma, conseguimos deixar a ideia explícita numa cadeia/sequência de operações atômicas que podem falhar ou não:
- extraimos os nomes dos arquivos para constantes
- decodificamos o base64 do pdf, caso dê erro, já retornamos
- criamos a lista de arquivos a serem incluídos nos zip
- fazemos a operação de criação do zip em memória, de forma atômica
- retornamos os dados necessários apra enviar o download no controller